Add transaction_timestamp, transaction_date, statement_timestamp, clo…

vivekmenezes opened the pull request:
…ck_timestamp.

Fix current_timestamp, current_date, now().

closes #2427

petermattis commented on sql/parser/builtins.go:
```diff
@@ -382,8 +382,8 @@ var builtins = map[string][]builtin{
 			types:      typeList{timestampType},
 			returnType: DummyInterval,
 			impure:     true,
```
I think this function and the other time functions (except `clock_timestamp`) are now pure.

vivekmenezes replied:
So purity is defined within the scope of a statement? I'd like to update the comment.

tamird replied:
Yeah, this is a bit ambiguous. If you look at how we use `impure` currently, we only care if the expression is a constant, and we only care about that for the purpose of expression simplification.

I think it would be fair to say that purity is statement-scoped, but the update comment should mention the specifics of how `impure` is accessed.

vivekmenezes replied:
I've updated the comment to reflect some of these thoughts. I prefer not mentioning how impure is used and the comments reflect only how a builtin writer would specify it.

petermattis commented on sql/testdata/builtin_function:
```diff
@@ -258,13 +258,75 @@ query error syntax error at or near "\("
+true
+
+query error syntax error at or near "\("
+SELECT clock_timestamp()
```
Is there a reason you want the function form to not work?

vivekmenezes replied:
No strong reason, we should support all functions like unique_id() without the parenthesis. It's nice to not have the parenthesis, making it look like a variable in SQL. I'll allow the variable use as a function.

vivekmenezes replied:
done

petermattis commented on sql/parser/sql.y:
```diff
@@ -3576,6 +3592,7 @@ reserved_keyword:
 | CASE
 | CAST
 | CHECK
+| CLOCK_TIMESTAMP
```
These are not reserved keywords in the SQL sense. I think you should be adding these "keywords" to the `type_func_name_keyword` list.

vivekmenezes replied:
done

petermattis commented on sql/executor.go:
```diff
@@ -72,6 +75,8 @@ func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
 			txn.SetSystemDBTrigger()
 		}
 		planMaker.txn = txn
+		now = time.Unix(0, 0).Add(time.Duration(planMaker.session.UnixTimestamp))
```
`proto.Timestamp.GoTime()` does this differently. I'm not sure if it matters, but it would be good be consistent. For that matter, it might be easier if `session.UnixTimestamp` were a `proto.Timestamp` so that you could just use the `GoTime()` method.

tamird replied:
:+1: assuming s/`proto.Timestamp`/`driver.Datum.Timestamp`/

vivekmenezes replied:
fixed

petermattis commented on sql/testdata/builtin_function:
```diff
@@ -258,13 +258,75 @@ query error syntax error at or near "\("
 
 query T
-SELECT now() - current_timestamp < interval '10s'
+SELECT now() - transaction_date::timestamp < interval '24h10s'
```
Other tests for `now()` exist in `testdata/datetime`. Perhaps put these new datetime function tests there as well.

vivekmenezes replied:
i can do that once you've approved this change. I'll move this code into the other file as a separate commit in this PR.

tamird commented on proto/data.pb.go:
```diff
@@ -546,7 +546,6 @@ func (m *NodeList) GetNodes() []int32 {
 // transaction is assigned a random priority. This priority will be
 // used to decide whether a transaction will be aborted during
 // contention.
-// TODO(vivek): Remove parts of Transaction that expose internals.
```
looks like you edited generated code.

vivekmenezes replied:
oops! fixed

vivekmenezes commented on sql/testdata/datetime:
```diff
@@ -72,3 +72,5 @@ query B
 ----
 true true
+
+
```
let me take this out

petermattis commented on sql/session.proto:
```diff
@@ -35,4 +35,7 @@ message Session {
   optional bool mutates_system_db = 4 [(gogoproto.nullable) = false, (gogoproto.customname) = "MutatesSystemDB"];
+  // Timestamp to be used by SQL in the above transaction. Note: this is not the
+  // transaction timestamp in proto.Transaction above.
+  optional proto.Timestamp unix_timestamp = 5 [(gogoproto.nullable) = false];
```
We usually name this field `timestamp`. I'm not sure why we need to highlight that it is a unix timestamp in the name. The comments in `proto.Timestamp` already indicate that it is based on the unix epoch.

tamird replied:
this is not the right type. We should use the driver's `Timestamp` here, not `proto`'s/

petermattis commented:
LGTM

tamird commented on sql/parser/sql.y:
```diff
@@ -2773,15 +2774,15 @@ func_expr_windowless:
   {
     $$ = &FuncExpr{Name: &QualifiedName{Base: Name($1)}}
   }
+| COLLATION FOR '(' a_expr ')' { unimplemented() }
```
nit: why move this?

tamird commented:
LGTM modulo the timestamp using the correct proto message type.

bdarnell commented on sql/executor.go:
```diff
@@ -72,6 +75,8 @@ func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
 			txn.SetSystemDBTrigger()
 		}
 		planMaker.txn = txn
+		now = planMaker.session.UnixTimestamp.GoTime()
```
It's weird that after this point `now` is not the current time. This should use a different local variable `txnTimestamp` (which defaults to `now`).

vivekmenezes replied:
sounds good

bdarnell commented on sql/parser/builtins.go:
```diff
@@ -399,15 +401,46 @@ var builtins = map[string][]builtin{
+
+	"statement_timestamp": {nowImpl},
+	"current_timestamp":   {nowImpl},
+	"now":                 {nowImpl},
```
In postgres, `now()` and `current_timestamp()` are equivalent to `transaction_timestamp()`, not `statement_timestamp()`.

bdarnell replied:
I don't feel strongly which way we go. It would be good, however, if functions like `current_date()` and `age()` were defined in terms of `nowImpl` to ensure that we're consistent.

vivekmenezes replied:
they all share e.StmtTimestamp to ensure consistency.

bdarnell replied:
Yes, they all share `e.StmtTimestamp`, but they all independently choose to use StmtTimestamp instead of TxnTimestamp. What I'm saying is that all of these should call into one place so that if we decide to make `now()` synonymous with `transaction_timestamp` there is only one place to change.

vivekmenezes replied:
They all call nowTimestamp()

bdarnell commented on sql/parser/builtins.go:
```diff
@@ -399,15 +401,46 @@ var builtins = map[string][]builtin{
+		},
+	},
+
+	"transaction_date": {
```
Does any other database have a `transaction_date()` function? I'm having a hard time seeing a use for it and it's always expressible as `cast(transaction_timestamp() as date)`.

vivekmenezes replied:
I'll take it out

bdarnell commented on sql/parser/sql.y:
```diff
@@ -2807,6 +2808,14 @@ func_expr_common_subexpr:
 | GREATEST '(' expr_list ')' { unimplemented() }
 | LEAST '(' expr_list ')' { unimplemented() }
 
+func_or_variable:
```
Only current_date and current_time have the func_or_variable treatment in postgres; the others are only functions. I'd prefer to minimize the number of names that have this special case.

vivekmenezes replied:
I'm in favor of removing the variable treatment for all these functions. The "modernized" approach does that.

bdarnell replied:
The variable form of `current_timestamp` is present in the SQL standard, and some databases forbid the use of parentheses with it (http://stackoverflow.com/questions/10084967/is-oracles-current-timestamp-function-really-a-function, although this information is suspect since it says that postgres disallows parens)

vivekmenezes replied:
we decided to kill the variable form for now since a functional form is consistent, and gels with a modern approach, and will punt changing this back to the standard variable form.

bdarnell commented on sql/testdata/builtin_function:
```diff
@@ -254,16 +254,70 @@ SELECT now() - current_date::timestamp < interval '24h10s'
+true
+
+query T
+SELECT current_timestamp() - statement_timestamp() = interval '0s'
```
It would be nice if we could have some tests where statement_timestamp and transaction_timestamp differ, but I don't see how to do that without introducing ugly test-only hooks.

petermattis replied:
We could add something to the logic-test framework (e.g. `sleep 10ms`).

tamird replied:
or we could do it in driver_test?

tamird commented on sql/executor.go:
```diff
@@ -54,10 +54,13 @@ func (e *Executor) SetNodeID(nodeID proto.NodeID) {
 // Execute the statement(s) in the given request and return a response.
 // On error, the returned integer is an HTTP error code.
 func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
+	txnTimestamp := time.Now()
```
this is a misnomer, this is the statement timestamp, not the transaction's.

bdarnell replied:
It's both: the statement timestamp and transaction timestamp are the same by default until we read a new value for txnTimestamp from the session Transaction. But I agree it would be clearer to have two separate variables.

vivekmenezes replied:
yeah they aught to be separate and created elsewhere.

bdarnell commented on rpc/codec/wire/.!8913!wire.pb.go:
```diff
@@ -0,0 +1,155 @@
+// Code generated by protoc-gen-gogo.
```
You've got some sort of garbage file here.

tamird commented on sql/executor.go:
```diff
@@ -83,6 +88,7 @@ func (e Executor) Execute(args driver.Request) (driver.Response, int, error) {
 	if planMaker.txn != nil {
 		planMaker.session.Txn = &planMaker.txn.Proto
 		planMaker.session.MutatesSystemDB = planMaker.txn.SystemDBTrigger()
+		planMaker.session.Timestamp = driver.Timestamp(txnTimestamp)
```
I think we need to put this timestamp on the `Txn` (which might mean we need a new proto message that wraps `proto.Transaction`). it's really not session state, and this is fragile as a result; if the transaction is committed all its associated state should be blown away.

bdarnell replied:
Why is that a problem?

tamird replied:
as discussed:

```sql
CREATE TABLE t (a timestamp);

BEGIN; INSERT INTO t VALUES (transaction_timestamp()); COMMIT; SELECT SLEEP (1); BEGIN; SELECT transaction_timestamp(), * FROM t; COMMIT;
```

vivekmenezes replied:
we talked and we decided to create a new message that embeds proto.Transaction and adds a sql transaction timestamp.

tamird commented on sql/executor.go:
```diff
@@ -54,10 +54,13 @@ func (e *Executor) SetNodeID(nodeID proto.NodeID) {
 		evalCtx: parser.EvalContext{
-			NodeID: e.nodeID,
+			NodeID:        e.nodeID,
+			StmtTimestamp: txnTimestamp,
```
by the way, this suffers a similar problem to the one described below. Concretely, this will produce the wrong output for this case:

```sql
INSERT INTO t VALUES (statement_timestamp()); SELECT SLEEP (1); SELECT statement_timestamp(), * FROM t;
```

vivekmenezes commented:
Hey, this PR has changed quite a bit that it's worth taking another look at it.

petermattis commented on sql/testdata/datetime:
```diff
@@ -72,3 +82,175 @@ query B
+# Check that the transaction_timestamp changes with each transaction.
+# We use, SELECT * FROM kv, to insert delays.
+statement ok
+BEGIN; INSERT INTO kv (k,v) VALUES ('b', transaction_timestamp()); SELECT * FROM kv; INSERT INTO kv (k,v) VALUES ('c', transaction_timestamp()); SELECT * FROM kv; INSERT INTO kv (k,v) VALUES ('d', current_timestamp()); SELECT * FROM KV; COMMIT; BEGIN; SELECT * FROM KV; INSERT INTO kv (k,v) VALUES ('e', transaction_timestamp()); COMMIT; 
```
FYI, this does not need to be one line. The logic-test code reads statements to execute up until the next blank line. That's how the `CREATE TABLE` at line 143 works.

vivekmenezes replied:
done

petermattis commented on storage/engine/.!9097!mvcc.pb.go:
```diff
@@ -0,0 +1,521 @@
+// Code generated by protoc-gen-gogo.
```
What is this file?

petermattis commented on util/.!7380!unresolved_addr.pb.go:
```diff
@@ -0,0 +1,172 @@
+// Code generated by protoc-gen-gogo.
```
Ditto. What is this file?

petermattis commented on sql/plan.go:
```diff
@@ -34,6 +35,15 @@ type planner struct {
 	evalCtx parser.EvalContext
 }
 
+func (p *planner) assignTxn(txn *client.Txn, timestamp time.Time) {
```
While this is an assignment, I think this would more typically be named `setTxn`.

tamird replied:
shouldn't `planner.txn` be a `Txn` rather than `client.Txn`?

tamird replied:
hm, perhaps not.

petermattis commented on sql/parser/sql.y:
```diff
@@ -3596,10 +3588,7 @@ reserved_keyword:
 | CONSTRAINT
 | CREATE
 | CURRENT_CATALOG
-| CURRENT_DATE
```
In general, you shouldn't be touching the `reserved_keyword` rule. Why was this change necessary? I'd like to see it reverted.

vivekmenezes replied:
reverted

petermattis commented on sql/parser/sql.y:
```diff
@@ -2774,14 +2774,6 @@ func_expr_windowless:
 // Special expressions that are considered to be functions.
 func_expr_common_subexpr:
   COLLATION FOR '(' a_expr ')' { unimplemented() }
-| CURRENT_DATE
```
I think you should revert this. The variable form of `CURRENT_{DATE,TIMESTAMP}` is part of the SQL standard.

vivekmenezes replied:
We decided to not go the standards route because it felt out of place and
is not part of the modern approach. We can always add in the standards
style later.

On Thu, Sep 24, 2015, 3:16 PM Peter Mattis <notifications@github.com> wrote:

> In sql/parser/sql.y
> <https://github.com/cockroachdb/cockroach/pull/2630#discussion_r40360285>:
>
> > @@ -2774,14 +2774,6 @@ func_expr_windowless:
> >  // Special expressions that are considered to be functions.
> >  func_expr_common_subexpr:
> >    COLLATION FOR '(' a_expr ')' { unimplemented() }
> > -| CURRENT_DATE
>
> I think you should revert this. The variable form of
> CURRENT_{DATE,TIMESTAMP} is part of the SQL standard.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2630/files#r40360285>.
>

tamird replied:
who is we?

petermattis replied:
Huh, did I miss that discussion? I think this is a mistake. And it's not like this a burden to support. FYI, your "modern approach" information is incorrect. We need to support `CURRENT_{TIMESTAMP,DATE}` without parentheses.

vivekmenezes replied:
Will put them both back.

tamird commented on sql/executor.go:
```diff
@@ -236,9 +234,9 @@ func (e Executor) execStmt(stmt parser.Statement, params parameters, planMaker *
 	// auto-transaction.
 	err := e.db.Txn(func(txn *client.Txn) error {
-		planMaker.txn = txn
+		planMaker.assignTxn(txn, timestamp)
```
i don't think it's correct for this to use `timestamp` as computed earlier

vivekmenezes replied:
Any particular reason why?

tamird replied:
well the clock continues to run - consider the case of an insert statement using transaction_timestamp: this implementation allow drift in the face of retries

vivekmenezes replied:
Very good point. I've fixed it.

vivekmenezes committed:
Add transaction_timestamp, statement_timestamp, and clock_timestamp.

Fix current_timestamp, current_date, now().

Fix count, max, and min for date, timestamp, and interval types.

Move tests for all date/timestamp/interval builtins to sql/testdata/datetime.

tamird commented on sql/parser/builtins.go:
```diff
@@ -399,15 +401,36 @@ var builtins = map[string][]builtin{
+			types:      typeList{},
+			returnType: DummyTimestamp,
+			fn: func(e EvalContext, args DTuple) (Datum, error) {
+				return DTimestamp{Time: e.TxnTimestamp}, nil
```
use `nowTimestamp()`

tamird replied:
err, disregard

tamird commented:
LGTM
